pkg/app/piped/deploysource/deploysource.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2020 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploysource
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+
+	"github.com/pipe-cd/pipe/pkg/config"
+	"github.com/pipe-cd/pipe/pkg/git"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+type DeploySource struct {
+	RepoDir                 string
+	AppDir                  string
+	RevisionName            string
+	Revision                string
+	DeploymentConfig        *config.Config
+	GenericDeploymentConfig config.GenericDeploymentSpec
+}
+
+type Provider interface {
+	Get(ctx context.Context, logWriter io.Writer) (*DeploySource, error)
+	GetReadOnly(ctx context.Context, logWriter io.Writer) (*DeploySource, error)
```

Comment:
Curious about why logWriter is given as an argument. Why don't you set it when calling `NewProvider`?

Reply:
Because the provider is initialized by the scheduler or planner and passed around stages to reuse the prepared data.
While each stage has a different LogPersister.

Reply:
Ah, now I turned out why it does, thanks:-D